### PR TITLE
Remove six

### PR DESCRIPTION
**Pull request by hramezani**
There is another possibility to remove the dependency to six.
we can move the required function of six like `ensure_str`, `ensure_binary`, ... and then remove `six.py`

### CODE REVIEW

File: src/urllib3/util/request.py
Comment (sethmlarson) [inline src/urllib3/util/request.py:82]: `b64encode` can only return ASCII so we can do `b64encode(x.encode("latin-1")).decode()`. Also we can use an `f-string here instead of string concat

File: test/test_filepost.py
Comment (sethmlarson) [inline test/test_filepost.py:7]: Instead of `LATIN_ENCODED_BOUNDARY` maybe `BOUNDARY_BYTES` is a more descriptive name?

File: test/test_response.py
Comment (sethmlarson) [inline test/test_response.py:803]: We should remove it and only do `.encode()` to not confuse future readers into thinking there's anything relevant about the "latin-1" encoding here.

File: test/with_dummyserver/test_connectionpool.py
Comment (sethmlarson) [inline test/with_dummyserver/test_connectionpool.py:702]: Same as above, remove latin-1 encoding.

**Review by sethmlarson — CHANGES_REQUESTED** [review]
Thanks for this! I left you some comments.

**Comment by hramezani** [comment]
@sethmlarson Thanks for the review. I addressed your comments.

What is your idea about my comment:

> we can move the required function of six like ensure_str, ensure_binary, ... and then remove six.py

If you agree with moving these function out of `six.py` please give me some hint about the file that we need to move these function to it.

**Comment by sethmlarson** [comment]
Definitely, maybe create a new module (`urllib3.util.util`??) since `urllib3.util` is unfortunately public.

**Comment by hramezani** [comment]
> Definitely, maybe create a new module (`urllib3.util.util`??) since `urllib3.util` is unfortunately public.

I will try to make a PR for it.

**Comment by sethmlarson** [comment]
Okay, so this PR is good to go then?

**Comment by hramezani** [comment]
we can merge this one and make another one for the six functions. Or have them all in this patch.
Which one do you prefer @sethmlarson?

**Comment by sethmlarson** [comment]
Let's bring everything into this PR, then we can label it "remove six"

File: test/test_filepost.py
Comment (sethmlarson) [inline test/test_filepost.py:7]: Can we remove the latin-1 here as well?

File: src/urllib3/util/util.py
Comment (sethmlarson) [inline src/urllib3/util/util.py:1]: Thoughts on using `to_str` and `to_bytes` as both shorter and more descriptive?

The following are optimized such that `to_X()` on type `X` is fast and we only pay the "not isinstance()" tax when it's not `X`. Also calling without any arguments is a lot faster than with any arguments so we should optimize for this case:

```python
from typing import Union, Optional


def to_bytes(x: Union[str, bytes], encoding: Optional[str] = None, errors: Optional[str] = None) -> bytes:
    if isinstance(x, bytes):
        return x
    elif not isinstance(x, str):
        raise TypeError(f"not expecting type {type(x).__name__}")
    elif encoding or errors:
        return x.encode(encoding, errors=errors)
    return x.encode()


def to_str(x: Union[str, bytes], encoding: Optional[str] = None, errors: Optional[str] = None) -> str:
    if isinstance(x, str):
        return x
    elif not isinstance(x, bytes):
        raise TypeError(f"not expecting type {type(x).__name__}")
    elif encoding or errors:
        return x.decode(encoding, errors=errors)
    return x.decode()
```

That way `to_bytes()` will encode as ASCII with errors=strict

File: src/urllib3/util/url.py
Comment (sethmlarson) [inline src/urllib3/util/url.py:213]: This call to `to_str()` would need an `encoding="utf-8"` if implemented like below.

File: src/urllib3/util/url.py
Comment (sethmlarson) [inline src/urllib3/util/url.py:399]: Idle thought: I feel like this isn't necessary anymore since we only accept URLs that are `str` for Python 3 so URL components can only end up as `str`.

**Review by sethmlarson — COMMENTED** [review]
A suggestion on name/implementation, would like your thoughts

File: src/urllib3/util/util.py
Comment (hramezani) [reply src/urllib3/util/util.py:1]: @sethmlarson Thanks for the suggestion.
I would suggest using default values for `encoding` and `errors` because with 
```python
    elif encoding or errors:
        return x.decode(encoding, errors=errors)
````
If we pass just `encoding`, the condition will be True and we will face `TypeError` (because the `error` is None)
`TypeError: decode() argument 2 must be str, not None`

Comment (sethmlarson) [reply src/urllib3/util/util.py:1]: Ah yes, perhaps `x.encode(encoding or "ascii", errors=errors or "strict")`. Trying to detect the case when we can simple `x.decode()` since it's >4x faster than calling with any arguments and encoding/decoding is a common operation for us.

File: src/urllib3/util/url.py
Comment (hramezani) [reply src/urllib3/util/url.py:213]: > would need an encoding="utf-8" if implemented like below.

Do you mean `encoding=encoding`? because the `_encode_invalid_chars` accepts an `encoding` param with default value to `"utf-8"`

Comment (sethmlarson) [reply src/urllib3/util/url.py:213]: Hm, that parameter has been doing basically nothing it seems. You can remove the parameter and at the end of the `_encode_invalid_chars()` change `encoded_component.decode(encoding)` to `.decode()` since percent-encoding implies the end result is ASCII.

File: src/urllib3/contrib/securetransport.py
Comment (sethmlarson) [inline src/urllib3/contrib/securetransport.py:870]: Let's expand the import to `from ..util.util import to_bytes` so it's similar to the rest of the usages

File: src/urllib3/contrib/pyopenssl.py
Comment (sethmlarson) [inline src/urllib3/contrib/pyopenssl.py:452]: nit: Let's expand the import to `from ..util.util import to_bytes` so it's similar to the rest of the usages

File: src/urllib3/util/url.py
Comment (sethmlarson) [inline src/urllib3/util/url.py:399]: :tada: This is so great!

File: src/urllib3/util/util.py
Comment (sethmlarson) [inline src/urllib3/util/util.py:4]: Since I wrote this code would be good to get another outside review (cc @pquentin @hodbn )

**Review by sethmlarson — DISMISSED** [review]
besides the two nits and the call for reviews I'm happy with this change, goodbye six!

**Comment by sethmlarson** [comment]
Oh also I wonder if there's some coverage loss because the `to_X()` `TypeError` case isn't covered. Maybe can throw in a couple test cases for our new utility methods? It's fine to mark `reraise` as not requiring code coverage since we lifted it from six.

File: src/urllib3/contrib/securetransport.py
Comment (hramezani) [reply src/urllib3/contrib/securetransport.py:870]: I get this error when I import it by `from ..util.util import to_bytes`

`Could not import SecureTransport: ModuleNotFoundError("No module named 'urllib3.packages.util'")`

File: src/urllib3/contrib/pyopenssl.py
Comment (hramezani) [reply src/urllib3/contrib/pyopenssl.py:452]: doc tests will fail with this change.

Maybe it is related to `from .. import util` that exists at the top of file

**Comment by hramezani** [comment]
@sethmlarson Thanks for the review.

I added two tests to cover `TypeError` in the newly adde to `to_X()` functions.
Also, I ignored coverage for `def reraise` by adding it to the `exclude_lines ` section of `.coveragerc` (not sure that I did it right)

**Review by pquentin — CHANGES_REQUESTED** [review]
Great work! I have a question about `to_bytes`

**Comment by pquentin** [comment]
Also, to_bytes and to_str are not fully covered as you can see here: https://codecov.io/gh/urllib3/urllib3/src/3d79bccbda1486de6945d0a084b782a43c77890e/src/urllib3/util/util.py

Capture d’écran 2020-11-22 à 21 17 36

File: src/urllib3/util/util.py
Comment (hramezani) [reply src/urllib3/util/util.py:13]: Since we want to call `x.encode()`, because it's faster (based on @sethmlarson [comment](https://github.com/urllib3/urllib3/pull/2078#discussion_r527957964). I think we can change it like this:

`return x.encode(encoding or "utf-8", errors=errors or "strict")`

**Comment by hramezani** [comment]
> Also, to_bytes and to_str are not fully covered as you can see here: https://codecov.io/gh/urllib3/urllib3/src/3d79bccbda1486de6945d0a084b782a43c77890e/src/urllib3/util/util.py
> 
> Capture d’écran 2020-11-22 à 21 17 36

I added some tests to complete the coverage.

File: src/urllib3/util/util.py
Comment (sethmlarson) [reply src/urllib3/util/util.py]: Wow, I didn't realize the default encoding for `str.encode()` was utf-8, thought it was ascii all this time. Thanks for bringing this to my attention!

In this case we will need to explicitly say `to_str("ascii")` in all the places I previously thought should be ASCII.

File: src/urllib3/contrib/pyopenssl.py
Comment (sethmlarson) [reply src/urllib3/contrib/pyopenssl.py:452]: This should be `to_bytes(p, "ascii")`

File: src/urllib3/contrib/securetransport.py
Comment (sethmlarson) [reply src/urllib3/contrib/securetransport.py:870]: `to_bytes(p, "ascii")`

File: src/urllib3/util/url.py
Comment (sethmlarson) [inline src/urllib3/util/url.py:213]: Can remove the `encoding=` here

File: src/urllib3/util/url.py
Comment (sethmlarson) [inline src/urllib3/util/url.py:289]: `to_str(x, "ascii")`

**Review by sethmlarson — CHANGES_REQUESTED** [review]
Sorry for the confusion, thank you @pquentin for correcting me. A few little comments left!

**Review by sethmlarson — DISMISSED** [review]
This looks good to me, looking for a final review from @pquentin now :)

File: test/test_util.py
Comment (pquentin) [inline test/test_util.py:786]:
```suggestion
            (b"test", {"encoding": "ascii"}, "test"),  # bytes input with ascii
```

File: test/test_util.py
Comment (pquentin) [inline test/test_util.py:795]:
```suggestion
        with pytest.raises(TypeError, match="not expecting type int"):
            to_str(1)
```

File: test/test_util.py
Comment (pquentin) [inline test/test_util.py:813]:
```suggestion
        with pytest.raises(TypeError, match="not expecting type int"):
            to_bytes(1)
```

File: .coveragerc
Comment (pquentin) [inline .coveragerc:25]: This file is for statements that occur multiple times, which is not the case here. And `reraise` is covered anyway, so we can remove that line.

```suggestion
```

File: src/urllib3/util/util.py
Comment (pquentin) [inline src/urllib3/util/util.py:13]: You can remove one branch here:

```suggestion
    return x.encode(encoding or "utf-8", errors=errors or "strict")
```

File: src/urllib3/util/util.py
Comment (pquentin) [inline src/urllib3/util/util.py:25]: You can remove one branch here:

```suggestion
    return x.decode(encoding or "utf-8", errors=errors or "strict")
```

File: test/test_util.py
Comment (pquentin) [inline test/test_util.py:804]:
```suggestion
            ("test", {"encoding": "ascii"}, b"test"),  # bytes input with ascii
```

**Review by pquentin — CHANGES_REQUESTED** [review]
Great work! We're nearly there, I only have cosmetic comments left.

**Comment by pquentin** [comment]
This now looks good to *me*, but contains code that I wrote, so I need review from either or you now!

File: src/urllib3/util/util.py
Comment (hramezani) [reply src/urllib3/util/util.py]: I think @sethmlarson suggested the branch in function to detect the case that we can call `x.encode()` because it's 4X faster. 
Take a look at this [comment](https://github.com/urllib3/urllib3/pull/2078#discussion_r527972377)

**Comment by pquentin** [comment]
Weird, but I seem to reproduce. I guess a comment would be best. I removed that part from the last commit, please review

**Review by sethmlarson — APPROVED** [review]
@pquentin LGTM, merge when ready!

**Comment by pquentin** [comment]
Closing/reopening as we ran into the coverage SQLite race condition again.

**Review by pquentin — APPROVED** [review]

**Comment by pquentin** [comment]
Goodbye six 👋 You served us well

**Comment by hramezani** [comment]
@sethmlarson @pquentin thanks for the review and your help.

**Pull request merged**
